### ISSUE TITLE
AuthGuard CanActivate returns user 'X' when not logged in, meaning AuthGuard doesn't protect pages

### ISSUE THREAD

**Issue by MichaelBurdett** [JsMobileDev/ionic-angular-firebase-authentication-starter#47]
Hello, I'm still learning the ways of Angular and Firebase so apologies in advance if my lack of understanding is the issue. I've GIT cloned and npm installed. Swapped the firebase credentals out for my own and tested logging in which worked.

I did notice that when not logged in, I was still taken to /home which is protected according to the `app-routing.module.ts` (Line 11). 

I've checked the auth-guard.ts file and found that the `user` variable being assigned `this.authService.getUser()`  (line 29) returns an 'X' object, making the next line, the `if (user)` statement (line 30), resolve true, even when there is no user. I can access /profile without logging in.

Is there something I should be changing my end?

Sorry again!

Thanks in advance,

Michael

![Screenshot 2022-05-11 at 20 49 36](https://user-images.githubusercontent.com/71025070/167935307-cba35288-5df0-4d89-8e3a-88a2e1e3aa1e.png)

**Comment by MichaelBurdett** [comment]
So after sleeping on it, I've come to a temporary fix (to anyone else who stumbles upon this) I replaced the canActivate function to one that uses the built in Firebase auth. Full auth.guard.ts below (Disclaimer, this might be a terrible idea but it's working as far as I can tell);

```
import {Injectable} from '@angular/core';
import {CanActivate, Router} from '@angular/router';
import {Auth, onAuthStateChanged} from '@angular/fire/auth';

@Injectable({
    providedIn: 'root'
})

export class AuthGuard implements CanActivate {
    constructor(
        private auth: Auth,
        private router: Router
    ) {}

    canActivate(): Promise<boolean> {
        return new Promise(async (resolve, reject) => {
            onAuthStateChanged(this.auth, async (user) => {
                if (user) {
                    console.log('User authenticated');
                    resolve(true);
                } else {
                    console.log('User redirected to login');
                    this.router.navigateByUrl('/login', { replaceUrl: true });
                    reject(false);
                }
            });
        });
    }
}
```

**Comment by javebratt** [comment]
That's an error on my part, I've been refactoring the template to lean more on RxJS instead of promises and probably missed that, I'll try to create a PR for this over the weekend, thanks for the catch :D

**Comment by MichaelBurdett** [comment]
Thanks dude! Appreciate this, I'm still trying to learn the wizardy of promises let alone RxJS - also wanted to say thanks for all the open-source things you've been doing - followed you since 2016 and the release of your e-book, although my career took me into different frameworks and languages, I'm trying to dive back into Ionic & Firebase and your resources are invaluable.

**Comment by javebratt** [comment]
Hey @MichaelBurdett sorry for the late reply, life has been busier than ever.

I pushed a new commit to fix the issue.

The problem is that the app was built fully using `async/await` and the `AuthenticationService` method was returning a promise, I refactored it to return an Observable but didn't update the `AuthGuard`.

Fixed it with the commit dae24a024d74eb57f3c715e0dbfaad306acdef0d
